### PR TITLE
fix various issues in extracting comments from original easyconfig file and including them again in dumped easyconfig

### LINKED ISSUE #3082: Misplaced comments after parsing and dumping of recipes
When Easybuild parses and dumps an easyconfig it can sometimes misplace or remove comments. Two examples are the following.

### `CDO-1.9.5-CrayGNU-19.09.eb`
Original easyconfig 
```
easyblock = 'ConfigureMake'

name = 'CDO'
version = '1.9.5'

homepage = 'https://code.zmaw.de/projects/cdo'
description = """
    CDO is a collection of command line Operators to manipulate and analyse
    Climate and NWP model Data.
"""

toolchain = {'name': 'CrayGNU', 'version': '19.09'}
toolchainopts = {'opt': True, 'pic': True}

# Files visible at https://code.mpimet.mpg.de/projects/cdo/files,
# however link is different for each version!
source_urls = ['https://code.mpimet.mpg.de/attachments/download/18264/']
sources = [SOURCELOWER_TAR_GZ]
checksums = [ '0c60f2c94dc5c76421ecf363153a5043' ]

dependencies = [
    ('cray-netcdf/4.6.3.1', EXTERNAL_MODULE),
    ('cray-hdf5/1.10.5.1', EXTERNAL_MODULE),
    # CDO: OpenMP support for compute intensive operators:
    # https://code.zmaw.de/projects/cdo/wiki/OpenMP_support
    # No MPI support
]


preconfigopts = 'export NC_CONFIG=`which nc-config` && '

configopts = "--with-hdf5=$HDF5_DIR --with-netcdf=$EBROOTNETCDFMINFORTRAN "

sanity_check_paths = {
    'files': ["bin/cdo"],
    'dirs': [],
}

```

And the same easyconfig after eb has parsed and dumped it. Comments inside dependencies are removed, except for the last line that goes after checksums.
```
easyblock = 'ConfigureMake'

name = 'CDO'
version = '1.9.5'

homepage = 'https://code.zmaw.de/projects/cdo'
description = """
    CDO is a collection of command line Operators to manipulate and analyse
    Climate and NWP model Data.
"""

toolchain = {'name': 'CrayGNU', 'version': '19.09'}
toolchainopts = {'opt': True, 'pic': True}

# Files visible at https://code.mpimet.mpg.de/projects/cdo/files,
# however link is different for each version!
source_urls = ['https://code.mpimet.mpg.de/attachments/download/18264/']
sources = [SOURCELOWER_TAR_GZ]
checksums = ['0c60f2c94dc5c76421ecf363153a5043']  # No MPI support

dependencies = [
    ('cray-netcdf/4.6.3.1', EXTERNAL_MODULE),
    ('cray-hdf5/1.10.5.1', EXTERNAL_MODULE),
]

preconfigopts = "export NC_CONFIG=`which nc-config` && "
configopts = "--with-hdf5=$HDF5_DIR --with-netcdf=$EBROOTNETCDFMINFORTRAN "


sanity_check_paths = {
    'files': ['bin/%(namelower)s'],
    'dirs': [],
}

moduleclass = 'data'
```

### `CrayCCE-19.09-classic.eb`
Original easyconfig
```
easyblock = 'CrayToolchain'

name = 'CrayCCE'
version = '19.09-classic'

homepage = 'https://pubs.cray.com/discover'
description = """Toolchain using Cray compiler wrapper, using PrgEnv-cray
(PE release: September 2019).\n"""

toolchain = SYSTEM

dependencies = [
    # PrgEnv version is not pinned, as Cray recommends to use the latest
    # (default) version
    ('PrgEnv-cray', EXTERNAL_MODULE),
    ('atp/2.1.3', EXTERNAL_MODULE),
    ('cce/9.0.2-classic', EXTERNAL_MODULE),
    ('modules/3.2.11.3', EXTERNAL_MODULE),
    ('cray-libsci/19.06.1', EXTERNAL_MODULE),
    ('cray-mpich/7.7.10', EXTERNAL_MODULE),
    ('craype/2.6.1', EXTERNAL_MODULE),
    ('pmi/5.0.14', EXTERNAL_MODULE),
]

# LD_LIBRARY_PATH is now updated by production.git/login/daint.footer

moduleclass = 'toolchain'
```
Similarly the comments inside dependencies are moved or removed. Dumped file:
```
easyblock = 'CrayToolchain'

name = 'CrayCCE'
version = '19.09-classic'

homepage = 'https://pubs.cray.com/discover'
description = """Toolchain using Cray compiler wrapper, using PrgEnv-cray
(PE release: September 2019).
"""

toolchain = {'name': 'system', 'version': 'system'}

dependencies = [
    ('PrgEnv-cray', EXTERNAL_MODULE),
    ('atp/2.1.3', EXTERNAL_MODULE),
    ('cce/9.0.2-classic', EXTERNAL_MODULE),
    ('modules/3.2.11.3', EXTERNAL_MODULE),
    ('cray-libsci/19.06.1', EXTERNAL_MODULE),
    ('cray-mpich/7.7.10', EXTERNAL_MODULE),
    ('craype/2.6.1', EXTERNAL_MODULE),
    ('pmi/5.0.14', EXTERNAL_MODULE),
]  # (default) version


# LD_LIBRARY_PATH is now updated by production.git/login/daint.footer
moduleclass = 'toolchain'
```

### PR DESCRIPTION
**Pull request by boegel**
fixes #3082

cc @ekouts @victorusu

### CODE REVIEW

File: easybuild/framework/easyconfig/format/one.py [hunk 1]
```diff
@@ -312,64 +340,134 @@ def extract_comments(self, rawtxt):
...
+            'tail': [],  # comment at the end of the easyconfig file
          }
 
-        rawlines = rawtxt.split('\n')
+        parsed_ec = self.get_config_dict()
 
-        # extract header first
-        while rawlines and rawlines[0].startswith('#'):
-            self.comments['header'].append(rawlines.pop(0))
+        param_def_regex = re.compile(r'^([a-z__0-9]+)\s*=')
```
Comment (akesandgren) [inline easybuild/framework/easyconfig/format/one.py]: Why the double "_" in the regex?

Comment (boegel) [reply easybuild/framework/easyconfig/format/one.py]: Ah, shouldn't be there. Doesn't matter much, but I'll clean it up.

File: test/framework/easyconfig.py [hunk 2]
```diff
@@ -1977,6 +1981,7 @@ def test_dump_comments(self):
             '# this is a header',
             '#',
             '# which may include empty comment lines',
+            '# weirdly indented lines',
```
Comment (akesandgren) [inline test/framework/easyconfig.py]: A bit of nitpicking, this should really keep the initial spaces, or?

Comment (boegel) [reply test/framework/easyconfig.py]: No, I disagree... The indent is wrong initially, and it's OK to have it properly indented in the dumped easyconfig.

Comment (akesandgren) [reply test/framework/easyconfig.py]: Ok

File: easybuild/framework/easyconfig/format/one.py [hunk 3]
```diff
@@ -150,12 +151,14 @@ def _reformat_line(self, param_name, param_val, outer=False, addlen=0):
 
         elif outer:
             # only reformat outer (iterable) values for (too) long lines (or for select parameters)
-            if isinstance(param_val, (list, tuple, dict)) and ((len(param_val) > 1 and line_too_long) or forced):
+            if isinstance(param_val, (list, tuple, dict)) and ((len(param_val) > 1 or line_too_long) or forced):
```
Comment (akesandgren) [inline easybuild/framework/easyconfig/format/one.py]: This one and -> or i don't fully understand. Could you explain it a bit?

Comment (boegel) [reply easybuild/framework/easyconfig/format/one.py]: I think this was a logic error...

It should always do one-element-per-line for lists with more than 1 element or very long lines, at least for "outer" iterables (so not necessarily for a list in a dict, like `sanity_check_paths`).

File: easybuild/framework/easyconfig/format/one.py [hunk 4]
```diff
@@ -312,64 +343,136 @@ def extract_comments(self, rawtxt):
...
+            """Grab more comment lines."""
+
+            comment_lines = []
+
+            while lines and (comment_regex.match(lines[0]) or whitespace_regex.match(lines[0])):
+                line = lines.pop(0)
+                _, actual_comment = split_on_comment_hash(line, param_key)
+                # prefix comment with '#' unless line was empty
+                if line:
+                    actual_comment = '# ' + actual_comment
```
Comment (akesandgren) [inline easybuild/framework/easyconfig/format/one.py]: If lines[0] is whitespace only, then wouldn't that end up as
# <line-with-only-witespace>

File: easybuild/framework/easyconfig/format/one.py [hunk 5]
```diff
@@ -312,64 +343,136 @@ def extract_comments(self, rawtxt):
...
+        rawlines = rawtxt.split('\n')
+
+        # extract header first (include empty lines too)
+        self.comments['header'] = grab_more_comment_lines(rawlines, None)
+
+        last_param_key = None
         while rawlines:
             rawline = rawlines.pop(0)
+
+            # keep track of last parameter definition we ran into
```
Comment (akesandgren) [inline easybuild/framework/easyconfig/format/one.py]: Is this comment really correct?
"Last parameter def we ran into" that's not quite what the code below does.
It finds the next param def, if any in the line, and keeps track of the last seen param def.
Not quite the same thing.

I think a better comment would be
"keep track of last seen parameter definition"

Comment (akesandgren) [reply easybuild/framework/easyconfig/format/one.py]: or possible add "including the current line"

File: test/framework/easyconfig.py [hunk 6]
```diff
@@ -1876,6 +1892,244 @@ def test_dump_comments(self):
...
+            "# trailing comment",
+            '',
+            "# with an empty line in between",
+            "# DONE!",
+        ])
+        ec = EasyConfig(None, rawtxt=rawtxt)
+
+        # check internal structure to keep track of comments
+        self.assertEqual(ec.parser._formatter.comments['above'], {
+           "    'https://anotherexample.com',": ['# annoying non-indented comment'],
```
Comment (akesandgren) [inline test/framework/easyconfig.py]: Shouldn't this one be in iterabove?

File: test/framework/easyconfig.py [hunk 7]
```diff
@@ -1876,6 +1892,244 @@ def test_dump_comments(self):
...
+            '\n'.join([
+                r'',
+                r"# the version doesn't matter much here",
+                r"version = ",
+            ]),
+            '\n'.join([
+                r'',
+                r"    # trailing comments in dependencies",
+                r"    # a bit weird, but it happens",
+                r"\]",
```
Comment (akesandgren) [inline test/framework/easyconfig.py]: Might as well add the inline comment to the test for the closing ].

File: easybuild/framework/easyconfig/format/one.py (hunk 4 above)
Comment (boegel) [reply easybuild/framework/easyconfig/format/one.py]: fixed by using `line.strip()`

File: test/framework/easyconfig.py (hunk 6 above)
Comment (boegel) [reply test/framework/easyconfig.py]: Yeah, good point, it should be. Fixed in da3532be8

File: test/framework/easyconfig.py (hunk 7 above)
Comment (boegel) [reply test/framework/easyconfig.py]: done in f7c14da

File: easybuild/framework/easyconfig/format/one.py (hunk 5 above)
Comment (boegel) [reply easybuild/framework/easyconfig/format/one.py]: comment clarified in 86ffb41

**Comment by akesandgren** [comment]
Going in, thanks @boegel!

**Pull request merged**